core/shared/src/main/scala/org/http4s/Query.scala

```diff
@@ -34,83 +34,257 @@ import java.nio.charset.StandardCharsets
+      case raw: Query.Raw =>
+        new Query.Parsed(raw.pairs ++ pairs)
+    }
 
-  def toVector: Vector[(String, Option[String])] = pairs
+  def toVector: Vector[(String, Option[String])] = this match {
+    case Query.Empty => Vector.empty
+    case parsed: Query.Parsed => parsed.pairs
+    case raw: Query.Raw => raw.pairs
+  }
 
-  def toList: List[(String, Option[String])] = toVector.toList
+  def toList: List[(String, Option[String])] = this match {
+    case Query.Empty => List.empty
+    case parsed: Query.Parsed => parsed.pairs.toList
+    case raw: Query.Raw => raw.pairs.toList
+  }
 
-  /** Render the Query as a `String`.
+  /** Map[String, String] representation of the [[Query]]
     *
-    * Pairs are separated by '&' and keys are separated from values by '='
+    * If multiple values exist for a key, the first is returned. If
+    * none exist, the empty `String` "" is returned.
     */
-  override def render(writer: Writer): writer.type = value.fold(
-    { pairs =>
+  lazy val params: Map[String, String] = this match {
+    case Query.Empty => Map.empty
+    case _: Query.Parsed | _: Query.Raw =>
+      multiParams.map { case (k, v) =>
+        k -> v.headOption.getOrElse("")
+      }
+  }
+
+  /** `Map[String, List[String]]` representation of the [[Query]]
+    *
+    * Params are represented as a `List[String]` and may be empty.
+    */
+  lazy val multiParams: Map[String, List[String]] = this match {
+    case Query.Empty => Map.empty
+    case _: Query.Parsed | _: Query.Raw =>
+      val pairs = toVector
+
+      if (pairs.isEmpty) Map.empty
+      else {
+        val m = mutable.Map.empty[String, ListBuffer[String]]
+        pairs.foreach {
+          case (k, None) => m.getOrElseUpdate(k, new ListBuffer)
+          case (k, Some(v)) => m.getOrElseUpdate(k, new ListBuffer) += v
+        }
+        m.view.mapValues(_.toList).toMap
+      }
+  }
+
+  override protected type Self = Query
+  override protected val query: Query = this
+  override protected def self: Self = this
+  override protected def replaceQuery(query: Query): Self = query
+}
+
+object Query {
+  case object Empty extends Query {
+    def pairs: Vector[KeyValue] = Vector.empty
+
+    override def render(writer: Writer): writer.type =
+      writer
+  }
+
+  final class Raw private[http4s] (value: String) extends Query {
+    private[this] var _pairs: Vector[KeyValue] = _
+
+    def pairs: Vector[KeyValue] = {
+      if (_pairs == null) {
+        _pairs = Query.parse(value)
+      }
+      _pairs
+    }
+
+    override def render(writer: Writer): writer.type =
+      writer.append(value)
+
+    override def equals(that: Any): Boolean =
+      that match {
+        case that: Query => that.toVector == toVector
+        case _ => false
+      }
+
+    override def hashCode: Int = 31 + pairs.##
+  }
+
+  final class Parsed private[http4s] (value: Vector[KeyValue]) extends Query {
+    def pairs: Vector[KeyValue] = value
```

Comment:
A nitpick maybe, but I think `value` is not necessary here:
```suggestion
  final class Parsed private[http4s] (override val pairs: Vector[KeyValue]) extends Query {
```

Reply:
Fair enough nitpick 👍🏻  It's a more concise implementation. Although I don't think this somehow impacts the runtime.